presets/ragengine/inference/inference.py

```diff
@@ -102,14 +102,91 @@ async def acomplete(self, prompt: str, formatted: bool = False, **kwargs: Any) -
+        except HTTPException as http_exc:
+            logger.error(f"HTTP exception during chat(): {http_exc.detail}")
+            raise http_exc
+        except Exception as e:
+            logger.error(f"Unexpected exception in chat(): {e}")
+            raise HTTPException(status_code=500, detail=f"An unexpected error occurred: {str(e)}")
 
-    async def _async_openai_complete(self, prompt: str, **kwargs: Any) -> CompletionResponse:
-        return await OpenAI(api_key=LLM_ACCESS_SECRET, **kwargs).acomplete(prompt)
+    @llm_chat_callback()
+    async def achat(
+        self,
+        messages: Sequence[ChatMessage],
+        **kwargs: Any,
+    ) -> ChatResponse:
+        """ Perform an asynchronous chat completion request. """
+        try:
+            base_model, base_max_len = self._get_default_model_info()
+            req = {
+                "model": self.get_param("model", base_model),
+                "max_tokens": self.get_param("max_tokens", base_max_len),
+                "messages": [{
+                    "role": message.role,
+                    "content": message.content if isinstance(message.content, str) else json.dumps(message.content)
+                } for message in messages if message.content is not None and message.content != ""
+                ],
+            }
 
-    async def _async_huggingface_remote_complete(self, prompt: str, **kwargs: Any) -> CompletionResponse:
-        data = {"prompt": prompt, **kwargs}
-        return await self._async_post_request(data, headers={"Authorization": f"Bearer {LLM_ACCESS_SECRET}", "Content-Type": "application/json"})
-    async def _async_custom_api_complete(self, prompt: str, **kwargs: Any) -> CompletionResponse:
+            # Add any additional parameters from self.params onto request
+            for key, value in self.params.items():
+                if key not in req:
+                    req[key] = value
+            
+            resp =  await self._async_post_request_raw(data=req, headers=DEFAULT_HEADERS)
+            return ChatResponse(
+                logprobs=resp.get("logprobs", None),
+                delta=resp.get("delta", None),
+                raw=resp,
+                message=ChatMessage(content=resp.get("choices", [{}])[0].get("message", {}).get("content", ""))
+            )
+        except HTTPException as http_exc:
+            logger.error(f"HTTP exception during achat(): {http_exc.detail}")
+            raise http_exc
+        except Exception as e:
+            logger.error(f"Unexpected exception in achat(): {e}")
+            raise HTTPException(status_code=500, detail=f"An unexpected error occurred: {str(e)}")
+        finally:
+            # Clear params after the completion is done
+            self.params = {}
+    
+    async def chat_completions_passthrough(self, chatCompletionsRequest: CompletionCreateParams, **kwargs: Any) -> ChatCompletionResponse:
+        try:
+            if "/chat/completions" not in LLM_INFERENCE_URL:
```

Comment:
This logic is a bit too coarse; we can refine it later. cc @bangqipropel

Reply:
Yeah sadly the LlamaIndex library doesnt give us all these options and response fields so I wanted the passthrough to be pretty straightforward. Whatever inference errors we get we pass back directly